wled00/espnow_broadcast.cpp

```diff
@@ -0,0 +1,249 @@
+}
+
+bool ESPNOWBroadcast::setup() {
+
+    auto err = esp_event_loop_create_default();
+    if ( ERR_OK != err && ESP_ERR_INVALID_STATE != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_STA_START, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_STA_STOP, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_AP_START, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+
+    return setupWiFi();
+}
+
+bool ESPNOWBroadcast::setupWiFi() {
+    Serial.println("ESPNOWBroadcast::setupWiFi()");
+
+    _state.exchange(STOPPED);
+
+    // To enable ESPNow, we need to be in WIFI_STA mode
+    if ( !WiFi.mode(WIFI_STA) ) {
+        Serial.println("WiFi.mode() failed");
+        return false;
+    }
+    // and not have the WiFi connect
+    // Calling discount with tigger an async Wifi Event
+    if ( !WiFi.disconnect(false, true) ) {
+        Serial.println("WiFi.disconnect() failed");
+        return false;
+    }
+
+    return true;
+}
+
+
+void ESPNOWBroadcast::loop(size_t maxMessagesToProcess /*= 1*/) {
+    switch (espnowBroadcast._state.load()) {
+        case ESPNOWBroadcast::STARTING:
+            // if WiFI is in starting state, actually stat ESPNow from our main task thread.
+            start();
+            break;
+        case ESPNOWBroadcast::STARTED: {
+            auto ndx = maxMessagesToProcess;
+            while(ndx-- > 0) {
+                size_t size = 0;
+                auto *msg = (QueuedNetworkMessage*)xRingbufferReceive(_networkRxMessages, &size, 0);
+                if (msg) {
+                    if(size == sizeof(*msg)) {
+                        auto callback = _rxCallback;
+                        while( *callback ) {
+                            (*callback)(msg->mac, msg->data, msg->len);
+                            callback++;
+                        }
+                    }
+                    vRingbufferReturnItem(_networkRxMessages, (void *)msg);
+                } else {
+                    break;
+                }
+            }
+            break;
+        }
+        default:
+            break;
+    }
+}
+
+ESPNOWBroadcast::err_t ESPNOWBroadcast::send(const uint8_t* msg, size_t len) {
+    static const uint8_t broadcast[] = BROADCAST_ADDR_ARRAY_INITIALIZER;
+    return static_cast<ESPNOWBroadcast::err_t>(esp_now_send(broadcast, msg, len));
+}
+
+void ESPNOWBroadcast::start() {
+
+    Serial.println("starting ESPNow");
+
+    if ( WiFi.mode(WIFI_STA) ) {
+        if ( WiFi.status() == WL_DISCONNECTED ) {
+            if (esp_wifi_start() == ESP_OK) {
+                if (esp_now_init() == ESP_OK) {
+                    if (esp_now_register_recv_cb(onESPNowRxCallback) == ESP_OK) {

```

Comment:
this seems like something we could just do at registration time rather than 
or does it have to be `init`ed for this to work?


... edit ... actually I just looked at https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/network/esp_now.html#_CPPv424esp_now_register_recv_cb17esp_now_recv_cb_t and apparently it does
So in that case, I would put an `unregister_cb` call into the de-init path.